pkg/cloudprovider/provider/nutanix/provider.go

```diff
@@ -0,0 +1,386 @@
+		if cleanupErr != nil {
+			return nil, fmt.Errorf("cleaning up failed with err %v after creation failed with err %v", cleanupErr, err)
+		}
+		return nil, err
+	}
+	return vm, nil
+}
+
+func (p *provider) create(machine *v1alpha1.Machine, userdata string) (instance.Instance, error) {
+	config, pc, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	client, err := GetClientSet(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct client: %v", err)
+	}
+
+	return createVM(client, machine.Spec.Name, *config, pc.OperatingSystem, userdata)
+}
+
+func (p *provider) Cleanup(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (bool, error) {
+	return p.cleanup(machine, data)
+}
+
+func (p *provider) cleanup(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (bool, error) {
+	config, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	client, err := GetClientSet(config)
+	if err != nil {
+		return false, fmt.Errorf("failed to construct client: %v", err)
+	}
+
+	project, err := getProjectByName(client, config.ProjectName)
+	if err != nil {
+		return false, fmt.Errorf("failed to get project: %v", err)
+	}
+
+	vm, err := getVMByName(client, machine.Name, *project.Metadata.UUID)
+	if err != nil {
+		if strings.Contains(err.Error(), entityNotFoundError) {
+			// VM is gone already
+			return true, nil
+		}
+
+		return false, fmt.Errorf("failed to get vm: %v", err)
+	}
+
+	// TODO: figure out if VM is already in deleting state
+
+	resp, err := client.Prism.V3.DeleteVM(*vm.Metadata.UUID)
+	taskID := resp.Status.ExecutionContext.TaskUUID.(string)
+
+	if err := waitForCompletion(client, taskID, time.Second*5, time.Minute*10); err != nil {
+		return false, fmt.Errorf("failed to wait for completion: %v", err)
+	}
+
+	return true, nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	config, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	client, err := GetClientSet(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct client: %v", err)
+	}
+
+	project, err := getProjectByName(client, config.ProjectName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get project: %v", err)
+	}
+
+	vm, err := getVMByName(client, machine.Name, *project.Metadata.UUID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get vm: %v", err)
+	}
+
+	if vm.Status == nil || vm.Status.Resources == nil || vm.Status.Resources.PowerState == nil {
+		return nil, errors.New("could not read vm power state")
+	}
+
+	var status instance.Status
+
+	switch *vm.Status.Resources.PowerState {
+	case "ON":
+		status = instance.StatusRunning
+	case "OFF":
+		status = instance.StatusCreating
```

Comment:
This is probably wrong but I wasn't sure what status to map to a VM that is off. Any input?